apps/sms/services/js/service_manager.js

```diff
@@ -0,0 +1,239 @@
+/* global App,
+          ConversationClient,
+          ConversationService,
+          MessagingClient,
+          MessagingService,
+          MozMobileConnectionsClient,
+          MozMobileConnectionsShim,
+          MozSettingsShim
+*/
+
+(function(exports) {
+  'use strict';
+
+  const debug = 1 ?
+    (arg1, ...args) => console.log(`[ServiceManager] ${arg1}`, ...args):
+    () => {};
+
+  const priv = {
+    upgradeKey: Symbol('upgradeKey')
+  };
+
+  const UPGRADE_KEY = 'upgrade-key';
+
+  // Allow old clients to gracefully finish all pending operations before
+  // we destroy them.
+  var ConversationServiceInitializer = {
```

Comment:
Can you explain a bit more what you have in mind? I just roughly combined related functionality in "initializers", but I'm sure there is much better option somewhere :)

Here "initialize()" is the method that initializes service (either via SharedWorker or directly if service is hosted in the same context) and "upgrade" is for "proper" upgrade of the window servce/client to sharedWorker/alwaysLowered-based  one.